solver/src/paraswap_solver.rs

```diff
@@ -1 +1,383 @@
+        Self {
+            settlement_contract,
+            token_info,
+            allowance_fetcher,
+            client: Box::new(DefaultParaswapApi::default()),
+        }
+    }
+}
+
+impl<F> std::fmt::Debug for ParaswapSolver<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str("ParaswapSolver")
+    }
+}
+
+/// Helper trait to mock the smart contract interaction
+#[cfg_attr(test, mockall::automock)]
+#[async_trait::async_trait]
+trait AllowanceFetching: Send + Sync {
+    async fn existing_allowance(&self, token: &ERC20, spender: &H160) -> Result<U256>;
+}
+
+#[async_trait::async_trait]
+impl AllowanceFetching for H160 {
+    async fn existing_allowance(&self, token: &ERC20, spender: &H160) -> Result<U256> {
+        Ok(token.allowance(*self, *spender).call().await?)
+    }
+}
+
+#[async_trait::async_trait]
+impl<F> SingleOrderSolving for ParaswapSolver<F>
+where
+    F: AllowanceFetching,
+{
+    async fn settle_order(&self, order: LimitOrder) -> Result<Settlement> {
+        let (amount, side) = match order.kind {
+            model::order::OrderKind::Buy => (order.buy_amount, Side::Buy),
+            model::order::OrderKind::Sell => (order.buy_amount, Side::Sell),
+        };
+        let token_infos = self
+            .token_info
+            .get_token_infos(&[order.sell_token, order.buy_token])
+            .await;
+        let decimals = |token: &H160| {
+            token_infos
+                .get(token)
+                .and_then(|info| info.decimals.map(usize::from))
+                .ok_or_else(|| anyhow!("decimals for token {:?} not found", token))
+        };
+
+        let price_query = PriceQuery {
+            from: order.sell_token,
+            to: order.buy_token,
+            from_decimals: decimals(&order.sell_token)?,
+            to_decimals: decimals(&order.buy_token)?,
+            amount,
+            side,
+        };
+        let price_response = self.client.price(price_query).await?;
+        ensure!(
+            satisfies_limit_price(&order, &price_response),
+            "Limit price not satisfied"
+        );
+
+        // 0.1% slippage
+        let src_amount_with_slippage = price_response
+            .src_amount
+            .checked_mul(1001.into())
+            .ok_or_else(|| anyhow!("Overflow during slippage computation"))?
+            / 1000;
+        let transaction_query = TransactionBuilderQuery {
+            src_token: order.sell_token,
+            dest_token: order.buy_token,
+            src_amount: src_amount_with_slippage,
+            dest_amount: price_response.dest_amount,
+            from_decimals: decimals(&order.sell_token)?,
+            to_decimals: decimals(&order.buy_token)?,
+            price_route: price_response.price_route,
+            user_address: self.settlement_contract.address(),
+            referrer: REFERRER.to_string(),
+        };
+        let transaction = self.client.transaction(transaction_query).await?;
```

Comment:
So their API service encodes the transaction for us, or is this something you had implemented in the previous PR?

Reply:
Correct, their API gives us the transaction calldata (same as 1Inch)